File: vyoma-ui/src/services/razorpayPrefill.ts
/** Razorpay prefill values must match strict patterns on mobile Safari. */

export type RazorpayPrefillInput = {
  name?: string | null
  email?: string | null
  contact?: string | null
}

export type RazorpayPrefill = {
  name?: string
  email?: string
  contact?: string
}

const EMAIL_RE = /^[^\s@]+@[^\s@]+\.[^\s@]+$/

/** E.164 India: +91 followed by 10 digits (Razorpay checkout expectation). */
export function formatIndiaContact(phone?: string | null): string | undefined {
  if (!phone?.trim()) return undefined
  const digits = phone.replace(/\D/g, '')
  if (digits.length < 10) return undefined
  const local = digits.length > 10 ? digits.slice(-10) : digits
  if (!/^\d{10}$/.test(local)) return undefined
  return `+91${local}`
}

export function buildRazorpayPrefill(input?: RazorpayPrefillInput): RazorpayPrefill | undefined {
  if (!input) return undefined

  const prefill: RazorpayPrefill = {}
  const name = input.name?.trim()
  if (name) prefill.name = name

  const email = input.email?.trim()
  if (email && EMAIL_RE.test(email)) prefill.email = email

  const contact = formatIndiaContact(input.contact)
  if (contact) prefill.contact = contact

  return Object.keys(prefill).length > 0 ? prefill : undefined
}

/** Receipt: max 40 chars, alphanumeric only (Razorpay validation). */
export function buildOrderReceipt(planId: string): string {
  const base = `vyoma${planId.replace(/[^a-z0-9]/gi, '')}${Date.now()}`
  return base.slice(0, 40)
}

File: lib/core/razorpay_prefill.dart
/// Razorpay checkout prefill — strict formats to avoid mobile validation errors.

String? formatIndiaContact(String? phone) {
  if (phone == null || phone.trim().isEmpty) return null;
  final digits = phone.replaceAll(RegExp(r'\D'), '');
  if (digits.length < 10) return null;
  final local = digits.length > 10 ? digits.substring(digits.length - 10) : digits;
  if (!RegExp(r'^\d{10}$').hasMatch(local)) return null;
  return '+91$local';
}

Map<String, String> buildRazorpayPrefill({
  String? name,
  String? email,
  String? contact,
}) {
  final prefill = <String, String>{};
  final trimmedName = name?.trim();
  if (trimmedName != null && trimmedName.isNotEmpty) {
    prefill['name'] = trimmedName;
  }
  final trimmedEmail = email?.trim();
  if (trimmedEmail != null &&
      trimmedEmail.isNotEmpty &&
      trimmedEmail.contains('@') &&
      !trimmedEmail.contains(' ')) {
    prefill['email'] = trimmedEmail;
  }
  final formattedContact = formatIndiaContact(contact);
  if (formattedContact != null) {
    prefill['contact'] = formattedContact;
  }
  return prefill;
}

String buildOrderReceipt(String planId) {
  final safe = planId.replaceAll(RegExp(r'[^a-zA-Z0-9]'), '');
  final receipt = 'vyoma$safe${DateTime.now().millisecondsSinceEpoch}';
  return receipt.length <= 40 ? receipt : receipt.substring(0, 40);
}
